README
PerlMaple version 0.04
======================

This is a very simple interface to Waterloo's Maple software via the OpenMaple
C interface.

INSTALLATION

Currently this software is only tested on Win32. To build this module
properly, you must first have Maple 9 or better installed on your 
system and append
the paths of B<maplec.h> and B<maplec.lib> in your Maple installation to the
environments LIB and INC respectively. Because this module use Maple's
C interface via L<Inline::C>.

Both F<maplec.h> and F<maplec.lib> (or maplec.a on *NIX systems?) are
provided by your Maple installation itself. A typical path of maplec.h
is "C:\Program Files\Maple 9\extern\include",
which should be appended to the INCLUDE environment. And a typical path
of maplec.lib is "C:\Program Files\Maple 9\bin.win", which should be
appended to the LIB environment. These paths may be different on your
machine but do depend on your Maple's version and location.

It may be similar on UNIX, but I haven't tried that.

It's known that this library's XS part won't work with Maple 7 on Win32.
Sigh.

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

Prerequisites

  Waterloo's Maple software (version 9 or higher)

DEPENDENCIES

  Carp
  Test::More
  Test::Deep

COPYRIGHT AND LICENCE

Copyright (C) 2005-2006 Agent Zhang

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
= Changes for PerlMaple 0.05 (rev 46) - Feb 18, 2006

== Documents

* confirmed that PerlMaple works both with Maple 9.02 and 10.00
* added a note to both README and PerlMaple.pm's POD for this
* used Kwid to reformat this file (Changes). Kwid++
* added some new TODOs to POD

= Changes for PerlMaple 0.04 (rev 43) - Feb 6, 2006

== Core Enhancements

* rewrote PerlMaple::Expression's constructor and made its instances 
  rather "lazy". That is to say, PerlMaple::Expression now won't
  construct a whole AST at startup time. It only evaluate a specific 
  part of the AST on the user's *actual* demand. This is really a big
  win on performance. Yay!
* added internal methods `init_type' and `init_ops' to PerlMaple::Expression.
  They're only called on demand of the users.
* tweaked the tests a bit to remain to pass

== Documents

* updated the POD to reflect recent changes in core
* added more words for the Maplesoft and its Maple product in the POD
  (to help Google find this stuff)
* moved the VERSION section of the POD in PerlMaple.pm to the front of
  the bloated SYNOPSIS section
* replaced the ugly `-E<gt>error' and things like that with more elegant
  `C<error>' or so.

= Changes for PerlMaple 0.03 (rev 38) - Feb 5, 2006

== Core Enhancements

* changed the PerlMaple error message from "PerlMaple error: XXX" to
  "PerlMaple: XXX". Indeed "errors" always frighten poeple arway.
* added undocumented and also preliminary support for the LogCmd attribute
* when Maple returns outputs of the type MAPLE_TEXT_WARNING, PerlMaple 
  will warn it out instead of return the string as a return value to the
  Perl programmer. (I'm sure there must be better solutions to this kind
  of things. :=))
* fixed a bug concerning zero number of ops
* added AUTOLOAD mechanism to PerlMaple::Expression so it now autoloads 
  Maple functions just as PerrlMaple does, which is a big win in syntax!
  Added POD docs and some tests for this new feature.
* fixed a bug which manifest itself when a PerlMaple::Expression object
  is contructed using a Maple assignment statement (say, a command 
  containing ":=").
* added more info (the Maple command given) to the carp/croak messages 
  issued by PerlMaple. It really eases debugging process.
* improved the operator overloading mechanism so that PerlMaple::Expression
  objects behave like simple Perl scalar variables. Yeah, simplicity is
  the key, no extra magic is really needed.

== Documents

* updated my email address in POD to agentzh@gmail.com
* added new TODO on facilities that imports complex perl data structure
  to Maple land
* added one more interesting example to PerlMaple's SYNOPSIS to 
  demonstrate the elegant AUTOLOAD feature of PerlMaple::Expression.
* added BUGS sections to both PerlMaple and PerlMaple::Expression's POD,
  mostly about the sharing of Maple evaluation contexts among PerlMaple
  objects.
* updated the outdated "Inline::C" comment to "XS".
* fixed the indented L<> linking in POD.
* added $VERSION to PerlMaple/Expression.pm
* Added TODO about making PerlMaple::Expression objects more "lazy" to 
  PerlMaple's POD.
* Removed duplicate code coverage report from PerlMaple::Expression's 
  POD.
* Confirmed that this lib won't work with Maple 7. Sigh.

= Changes for PerlMaple 0.02 (rev 11) - Dec 19, 2005

== Core Enhancements

* On the XS side, ignored MAPLE_TEXT_STATUS output so that
  PerlMaple won't return silly things like
  "bytes used=1963, alloc=719, time=0.44".
* Added AUTOLOAD mechanism so that any Maple functions or procedures
  all automagically become the methods of PerlMaple objects;
* Renamed ->eval method to ->eval_cmd so that ->eval is exactly the
  Maple's counterpart.
* ->eval is not backward-compatible.
* Added optional options PrintError and RaiseError to the constructor
* Added attribute methods PrintError and RaiseError
* Fixed a bug concerning multiple PerlMaple instances creation
* Added corresponding tests.
* Added the PerlMaple::Expression class to represents a Maple expression ASTs.
* PerlMaple::Expression can now build full-featured Perl AST from arbitrary
  Maple expressions. So you can manipulate Maple expressions via data
  structures directly, without the burden of parsing Maple code yourself!
* PerlMaple::Expression provides several useful mehods, such as type, expr,
  and ops.
* Added method to_ast to the PerlMaple class, which is construct a
  PerlMaple::Expression object from the given Maple expression.
* Added ReturnAST attribute to the PerlMaple class (default off).
* PerlMaple's AUTOLOADed methods (shortcuts for Maple functions) and the
  eval_cmd method will return a PerlMaple::Expression object automagically
  if you turn on the ReturnAST attribute.

== Documents

* Added SVN link 'https://svn.berlios.de/svnroot/repos/win32maple'
  to the POD doc.
* Emphasized that Maple's maplec.h and maplec.lib
  must be in INC and LIB, repectively. Most of the test failure
  reports are concerning this issue.
* Added many, many POD docs and many, many tests for the vast variety of
  new features.
* Added code coverage testing results to POD docs. (Thanks to Devel::Cover.)

= Changes for PerlMaple 0.01 (rev 1) - Nov 15, 2005

* original version; created by h2xs 1.23
* basic functionalities implemented


lib/PerlMaple.pm
#: PerlMaple.pm
#: implementation for the PerlMaple class
#: Copyright (c) 2005-2006 Agent Zhang
#: 2005-11-14 2006-02-18

package PerlMaple;

use 5.006;
use strict;
use warnings;

use PerlMaple::Expression;
use Carp qw(carp croak);
use vars qw( $AUTOLOAD );

our $VERSION = '0.05';
my $Started = 0;

require XSLoader;
XSLoader::load('PerlMaple', $VERSION);

sub new {
    my $class = shift;
    my %opts = @_;
    if (!exists $opts{PrintError}) {
        $opts{PrintError} = 1;
    }
    if ($Started or maple_start()) {
        $Started = 1;
        return bless \%opts, $class;
    }
    return undef;
}

sub eval_cmd {
    my ($self, $exp) = @_;
    $exp =~ s/[\s\n\r]+$//s;
    #if ($exp !~ /[;:]$/) {
    #    $exp .= ';';
    #}
    if ($self->{LogCmd}) {
        warn "$exp\n";
    }
    maple_eval($exp);
    if (maple_success()) {
        my $res = maple_result();
        return
            $self->{ReturnAST} ? $self->to_ast($res) : $res;
    }
    if ($self->{PrintError}) {
        carp "PerlMaple: ", $self->error,
            " when evaluating \"$exp\"";
    }
    if ($self->{RaiseError}) {
        croak "PerlMaple: ", $self->error,
            " when evaluating \"$exp\"";
    }
    return undef;
}

sub PrintError {
    my $self = shift;
    if (@_) {
        $self->{PrintError} = shift;
    } else {
        return $self->{PrintError};
    }
}

sub RaiseError {
    my $self = shift;
    if (@_) {
        $self->{RaiseError} = shift;
    } else {
        return $self->{RaiseError};
    }
}

sub ReturnAST {
    my $self = shift;
    if (@_) {
        $self->{ReturnAST} = shift;
    } else {
        return $self->{ReturnAST};
    }
}

sub error {
    if (!maple_success()) {
        return maple_error();
    }
    return undef;
}

sub to_ast {
    shift;
    return PerlMaple::Expression->new(@_);
}

sub AUTOLOAD {
    my $self = shift;
    my $method = $AUTOLOAD;
    #warn "$method";

    $method =~ s/.*:://;
    return if $method eq 'DESTROY';

    my $args = join(',', @_);
    $args =~ s/,\s*$//g;
    return $self->eval_cmd("$method($args);");
}

1;

__END__

=head1 NAME

PerlMaple - Perl binding for Maplesoft's Maple mathematical package

=head1 VERSION

This document describes PerlMaple 0.05 released on February 18, 2006.

=head1 SYNOPSIS

  use PerlMaple;

  $maple = PerlMaple->new or
      die $maple->error;
  $ans = $maple->eval_cmd('int(2*x^3,x);');
  defined $ans or die $maple->error;

  $maple->eval_cmd(<<'.');
  lc := proc( s, u, t, v )
           description "form a linear combination of the arguments";
           s * u + t * v
  end proc;
  .

  $maple = PerlMaple->new(
    PrintError => 0,
    RaiseError => 1,
  );

  print $maple->eval('solve(x^2+2*x+4, x)');  # got: -1+I*3^(1/2), -1-I*3^(1/2)
  print $maple->solve('x^2+2*x+4', 'x');  # ditto
  print $maple->eval('int(2*x^3,x)', 'x=2');  # got: 8s
  print $maple->int('2*x^3', 'x=0..2');  # ditto

  $maple->PrintError(1);
  $maple->RaiseError(0);

  print $maple->diff('2*x^3', 'x'); # got: 6*x^2
  print $maple->ifactor('1000!');  # will get a lot of stuff...

  # Advanced usage (manipulating Maple ASTs directly):
  $ast = $maple->to_ast('[1,2,3]');
  foreach ($ast->ops) {
      print;  # got 1, 2, and 3 respectively
  }

  # Get eval_cmd and other AUTOLOADed Maple function
  # to return ASTs automagically:
  $maple->ReturnAST(1);
  $ast = $maple->solve('x^2+x=0', 'x');
  if ($ast and $ast->type('exprseq')) {
    foreach ($ast->ops) {
        push @roots, $_;
    }
  }
  print "@roots";  # got: 0 -1

  $maple->ReturnAST(1);
  $res = $maple->solve('{x+y=2,x-y=3}', '{x,y}');
  if ($res->type('set')) {
      $x = first { $_->lhs eq 'x' } $res->ops;
      print $x->rhs;  # 5/2
  }

=head1 DESCRIPTION

Maple is a great tool for solving mathematical problems and creating
interactive technical applications provided by Maplesoft. Its power of 
symbolic calculation is extremely impressive.

This is a simple (but already powerful) Perl binding for Maple's
OpenMaple C interface. To try out this binding, you have to 
purchase and install the Maple software first:

L<http://www.maplesoft.com>.

The Maple software is *not* free, sigh, unlike this CPAN distribution.

=head1 INSTALLATION

Currently this software is only tested on Win32 against Maple 9.01 and
Maple 10.00, but it's believed to work with other versions higher than
9. If you get it work with a specific version of Maple, please send me
a mail and let me know. :=)

To build this module properly, you must first have Maple 9 or better 
installed on your system and append
the paths of B<maplec.h> and B<maplec.lib> in your Maple installation to the
environments LIB and INC respectively. Because this module use Maple's
C interface (OpenMaple) via XS.

Both F<maplec.h> and F<maplec.lib> (or maplec.a on *NIX systems?) are
provided by your Maple installation itself. A typical path of maplec.h
is "C:\Program Files\Maple 9\extern\include",
which should be appended to the INCLUDE environment. And a typical path
of maplec.lib is "C:\Program Files\Maple 9\bin.win", which should be
appended to the LIB environment. These paths may be different on your
machine but do depend on your Maple's version and location.

It may be similar on UNIX, but I haven't tried that.

It's known that this library's XS part won't work with Maple 7 on Win32.
Sigh.

=head2 EXPORT

None by default.

=head1 METHODS

PerlMaple uses AUTOLOAD mechanism so that any Maple functions are also valid
methods of your PerlMaple object, even those Maple procedures defined by
yourself.

When the ReturnAST attribute is on, all AUTOLOADed methods, along with the
C<eval_cmd> method will return a
PerlMaple::Expression object constructed from the resulting expression.
Because there's a cost involved in constructing an AST from the given Maple
expression, so the ReturnAST attribute is off by default.

B<WARNING:> The C<eval> method is now incompatible with that of the first
release (0.01). It is AUTOLOADed as the Maple counterpart. So you have to turn to
C<eval_cmd> when Maple's eval function can't fit your needs. Sorry.

=over

=item $obj->new()

=item $obj->new(RaiseError => 0, PrintError => 1, ReturnAST => 0)

Class constructor. It starts a Maple session if it does not exist. It should
be noting that although you're able to create more than one PerlMaple objects,
all these maple objects share the same Maple session. So the context of each
PerlMaple objects may be corrupted intentionally. If any error occurs, this
method will return undef value, and set the internal error buffer which you can
read by the C<error> method.

This method also accepts two optional named arguments. When RaiseError is
set true, the constructor sets the RaiseError attribute of the new object
internally. And it is the same for the PrintError and ReturnAST attributes.

=item $obj->eval_cmd($command)

This method may be the most important one for the implementation of this class.
It evaluates the
command stored in the argument, and returns a string containing the result
when the ReturnAST attribute is off. (It's off by default.)
If an error occurs, it will return undef, and set the internal error buffer
which you can read by the C<error> method.

Frankly speaking, most of the time you can use C<eval> method instead of
invoking this method directly. However, this method is a bit faster,
because any AUTOLOADed method is invoked this one internally. Moreover, there
exists something that can only be eval'ed properly by C<eval_cmd>.
Here is a small example:

    $maple->eval_cmd(<<'.');
    lc := proc( s, u, t, v )
             description "form a linear combination of the arguments";
             s * u + t * v
    end proc;
    .

If you use C<eval> instead, you will get the following error message:

    `:=` unexpected

That's because "eval" is a normal Maple function and hence you can't use
assignment statement as the argument.

When the ReturnAST attribute is on, this method will return a
PerlMaple::Expression object constructed from the expression returned by
Maple automatically.

=item $obj->to_ast($maple_expr, ?$verified)

This method is a shortcut for constructing a PerlMaple::Expression
instance. For more information on the PerlMaple::Expression class
and the second optional argument of this method, please turn to
L<PerlMaple::Expression>.

Here is a quick demo to illustrate the power of the PerlMaple::Expression
class (you can get many more in the doc for L<PerlMaple::Expression>.

    $ast = $maple->to_ast('[1,2,3]');
    my @list;
    foreach ($ast->ops) {
        push @list, $_->expr;
    }
    # now @list contains numbers 1, 2, and 3.

=item $obj->error()

It returns the error message issued by the Maple kernel.

=back

=head1 ATTRIBUTES

All the attributes specified below can be set by passing name-value pairs to
the C<new> method.

=over

=item $obj->PrintError()

=item $obj->PrintError($new_value)

The PrintError attribute can be used to force errors to generate warnings
(using Carp::carp) in addition to returning error codes in the normal way. When
set ``on'' (say, a true value), any method which results in an error
occurring will cause the PerlMaple to effectively do a
C<carp("PerlMaple error: ", $self->error, " when evaluating \"$exp\"";)>.

By default, the constructor C<new> sets PrintError ``on''.

=item $obj->RaiseError()

=item $obj->RaiseError($new_value)

The RaiseError attribute can be used to force errors to raise exceptions
rather than simply return error codes in the normal way. It is ``off''
(say, a false value in Perl) by default. When set ``on'', any method
which results in an error will cause the PerlMaple to effectively do a
C<croak("PerlMaple error: ", $self->error, " when evaluating \"$exp\"";)>.

If you turn RaiseError on then you'd normally turn PrintError off.
If PrintError is also on, then the PrintError is done first (naturally).

=item $obj->ReturnAST()

=item $obj->ReturnAST($new_value)

The ReturnAST attribute can be used to force the C<eval_cmd> method
and hence all AUTOLOADed Maple functions to return a Abstract
Syntactic Tree (AST). Because there is a cost to evaluate the C<to_ast>
method every time, so this attribute is off by default.

=back

=head1 INTERNAL FUNCTIONS

=over

=item maple_error

Returns the raw error message issued by the Maple kernel.

=item maple_eval

Raw XS C method that evaluates any Maple commands.

=item maple_result

Returns the raw output of the Maple kernel.

=item maple_start

Starts the global Maple kernel. (No that all PerlMaple objects share
the same Maple engine, so preventing context clashes is the
duty of the user.

=item maple_success

Indicates whether the last Maple evaluation is successful.

=back

=head1 CODE COVERAGE

I use L<Devel::Cover> to test the code coverage of my tests, below is the
L<Devel::Cover> report on this module's test suite (version 0.05):

    ---------------------------- ------ ------ ------ ------ ------ ------ ------
    File                           stmt   bran   cond    sub    pod   time  total
    ---------------------------- ------ ------ ------ ------ ------ ------ ------
    blib/lib/PerlMaple.pm          93.7   83.3   66.7  100.0  100.0  100.0   91.9
    ...b/PerlMaple/Expression.pm   98.7   88.9   66.7  100.0  100.0    0.0   94.6
    Total                          96.5   86.7   66.7  100.0  100.0  100.0   93.4
    ---------------------------- ------ ------ ------ ------ ------ ------ ------

=head1 BUGS

=over

At this point, every PerlMaple objects are merely a bunch of options (e.g.
PrintError, RaiseError, and ReturnAST). They actually share the same Maple
engine for performance reasons. So the evaluating contexts for each
PerlMaple objects may corrupt together. So please don't store
permanent values in Maple variables, avoid reference Maple special variables,
like %, %%, and don't trust the Maple contexts. Since they may be changed
accidentally by other PerlMaple objects, even those used by
PerlMaple::Expression instances, in the duration.

=back

=head1 TODO

=over

=item *

Add facilities to ease importing of Perl complex data structures to Maple
environment, i.e. a list of lists of numbers.

=item *

Port this binding to Linux Maple users and (hopefully) to MATLAB users as well.

=item *

Add the C<LogCommand> attribute so that PerlMaple will log down all the Maple 
commands sent to the Maple engine in disk files.

=item *

Added PerlMaple::Cookbook and PerlMaple::Tutorial to the distribution.
Programming Maple in Perl is extremely interesting and useful, so I
believe these docs are really worth my while.

=back

=head1 REPOSITORY

You can always get the latest version from the following SVN
repository:

L<https://svn.berlios.de/svnroot/repos/win32maple>

If you want a committer bit, please let me know.

=head1 SEE ALSO

L<PerlMaple::Expression>,
L<http://www.maplesoft.com>.

=head1 AUTHOR

Agent Zhang, E<lt>agentzh@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005-2006 Agent Zhang

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
